pkg/credentialprovider/gcp/metadata.go

```diff
@@ -91,210 +36,25 @@ func init() {
 		Timeout:   metadataHTTPClientTimeout,
 	}
 	credentialprovider.RegisterCredentialProvider("google-dockercfg",
-		&credentialprovider.CachingDockerConfigProvider{
-			Provider: &dockerConfigKeyProvider{
-				metadataProvider{Client: httpClient},
+		&credentialconfig.CachingDockerConfigProvider{
+			Provider: &gcpcredential.DockerConfigKeyProvider{
+				MetadataProvider: gcpcredential.MetadataProvider{Client: httpClient},
 			},
 			Lifetime: 60 * time.Second,
 		})
 
 	credentialprovider.RegisterCredentialProvider("google-dockercfg-url",
-		&credentialprovider.CachingDockerConfigProvider{
-			Provider: &dockerConfigURLKeyProvider{
-				metadataProvider{Client: httpClient},
+		&credentialconfig.CachingDockerConfigProvider{
+			Provider: &gcpcredential.DockerConfigURLKeyProvider{
+				MetadataProvider: gcpcredential.MetadataProvider{Client: httpClient},
 			},
 			Lifetime: 60 * time.Second,
 		})
 
 	credentialprovider.RegisterCredentialProvider("google-container-registry",
 		// Never cache this.  The access token is already
 		// cached by the metadata service.
-		&containerRegistryProvider{
-			metadataProvider{Client: httpClient},
+		&gcpcredential.ContainerRegistryProvider{
+			MetadataProvider: gcpcredential.MetadataProvider{Client: httpClient},
 		})
 }
-
-// Returns true if it finds a local GCE VM.
-// Looks at a product file that is an undocumented API.
-func onGCEVM() bool {
-	var name string
-
-	if runtime.GOOS == "windows" {
-		data, err := exec.Command("wmic", "computersystem", "get", "model").Output()
-		if err != nil {
-			return false
-		}
-		fields := strings.Split(strings.TrimSpace(string(data)), "\r\n")
-		if len(fields) != 2 {
-			klog.V(2).Infof("Received unexpected value retrieving system model: %q", string(data))
-			return false
-		}
-		name = fields[1]
-	} else {
-		data, err := ioutil.ReadFile(gceProductNameFile)
-		if err != nil {
-			klog.V(2).Infof("Error while reading product_name: %v", err)
-			return false
-		}
-		name = strings.TrimSpace(string(data))
-	}
-	return name == "Google" || name == "Google Compute Engine"
-}
-
-// Enabled implements DockerConfigProvider for all of the Google implementations.
-func (g *metadataProvider) Enabled() bool {
-	return onGCEVM()
-}
-
-// Provide implements DockerConfigProvider
-func (g *dockerConfigKeyProvider) Provide(image string) credentialprovider.DockerConfig {
```

Comment:
Alternative approach:

Add a helper function in `k8s.io/legacy-cloud-providers/gce` that covers this exact behavior, and instead of returning `credentialprovider.DockerConfig`, it turns a copy of DockeConfig local to the gce credential provider package. Then this call can be updated to call `Provide` from the gce provider and translate DockerConfig from GCE to the local DockerConfig type.

Reply:
Resolving this for the same reason given [here](https://github.com/kubernetes/kubernetes/pull/95775#discussion_r522529242) - but again, everyone should feel free to bring up another conversation if they see fit to do so.